pkg/api/customization/workload/workload.go

```diff
@@ -36,33 +38,108 @@ func (a ActionWrapper) ActionHandler(actionName string, action *types.Action, ap
+			return httperror.NewAPIError(httperror.ServerError, fmt.Sprintf("Error updating workload %s by %s : %s", deployment.ID, actionName, err.Error()))
 		}
 
-	case "pause", "resume":
-		data, err := convert.EncodeToMap(deployment)
-		if err == nil {
-			values.PutValue(data, !deployment.DeploymentConfig.Paused, "deploymentConfig", "paused")
-			err = update(apiContext, data, deployment.ID)
+	case "pause":
+		if deployment.Paused {
+			return httperror.NewAPIError(httperror.InvalidAction, fmt.Sprintf("Deployment %s already paused", deployment.ID))
 		}
-		if err != nil {
-			return httperror.NewAPIError(httperror.ServerError, fmt.Sprintf("Error updating workload %s by %s", deployment.ID, actionName))
+		return updatePause(apiContext, true, deployment, "pause")
+
+	case "resume":
+		if !deployment.Paused {
+			return httperror.NewAPIError(httperror.InvalidAction, fmt.Sprintf("Pause deployment %s before resume", deployment.ID))
+		}
+		return updatePause(apiContext, false, deployment, "resume")
+	}
+	return nil
+}
+
+func fetchRevisionFor(apiContext *types.APIContext, rollbackInput *projectclient.DeploymentRollbackInput, namespace string, name string, currRevision string) string {
+	rollbackTo := rollbackInput.ReplicaSetID
+	if rollbackTo == "" {
+		revisionNum, _ := convert.ToNumber(currRevision)
+		return convert.ToString(revisionNum - 1)
+
+	}
+	data := getRevisions(apiContext, namespace, name, rollbackTo)
+	if len(data) > 0 {
+		return convert.ToString(values.GetValueN(data[0], "workloadAnnotations", "deployment.kubernetes.io/revision"))
+	}
+	return ""
+}
+
+func getRevisions(apiContext *types.APIContext, namespace string, name string, requestedID string) []map[string]interface{} {
+	data, replicaSets := []map[string]interface{}{}, []map[string]interface{}{}
+	options := map[string]string{"hidden": "true"}
+	conditions := []*types.QueryCondition{
+		types.NewConditionFromString("namespaceId", types.ModifierEQ, []string{namespace}...),
+	}
+
+	if err := access.List(apiContext, &projectschema.Version, projectclient.ReplicaSetType, &types.QueryOptions{Options: options, Conditions: conditions}, &replicaSets); err == nil {
+		for _, replicaSet := range replicaSets {
+			replicaSetID := convert.ToString(replicaSet["id"])
+			if requestedID != "" && requestedID != replicaSetID {
+				continue
+			}
+			ownerReferences := convert.ToMapSlice(replicaSet["ownerReferences"])
+			for _, ownerReference := range ownerReferences {
+				kind := convert.ToString(ownerReference["kind"])
+				ownerName := convert.ToString(ownerReference["name"])
+				if kind == "Deployment" && name == ownerName {
+					data = append(data, replicaSet)
+					continue
+				}
+			}
 		}
 	}
+	return data
+}
+
+func updatePause(apiContext *types.APIContext, value bool, deployment projectclient.Workload, actionName string) error {
+	data, err := convert.EncodeToMap(deployment)
```

Comment:
can we just check if err != nil, and return right away? Then there is no need for an extra if

Reply:
wanted to avoid same```if err!=nil, return server error``` twice in a small function, so this way returning once if any one of errs is not nil, is that ok?